Mathlib/FieldTheory/Perfect.lean

```diff
@@ -111,14 +111,50 @@ theorem not_irreducible_expand (f : R[X]) : ¬ Irreducible (expand R p f) := by
   rw [polynomial_expand_eq]
   exact not_irreducible_pow hp.out.ne_one
 
-instance (S : Type*) [CommSemiring S] [CharP S p] [PerfectRing S p] :
+instance instPerfectRingProd (S : Type*) [CommSemiring S] [CharP S p] [PerfectRing S p] :
     PerfectRing (R × S) p := by
   constructor
   have : frobenius (R × S) p = Prod.map (frobenius R p) (frobenius S p) := by
     ext <;> simp [frobenius_def]
   rw [this]
   exact Bijective.Prod_map (bijective_frobenius R p) (bijective_frobenius S p)
 
+namespace Polynomial
+
+open scoped Classical
+
+variable {R : Type*} [CommRing R] [IsDomain R]
+  (p : ℕ) [Fact p.Prime] [CharP R p] [PerfectRing R p] (f : R[X])
+
+/-- If `f` is a polynomial over a perfect integral domain `R` of characteristic `p`, then there is
+a bijection from the set of roots of `Polynomial.expand R p f` to the set of roots of `f`.
+It's given by `x ↦ x ^ p`, see `rootsExpandEquivRoots_apply`. -/
+noncomputable def rootsExpandEquivRoots : (expand R p f).roots.toFinset ≃ f.roots.toFinset :=
+  ((frobeniusEquiv R p).image _).trans <| Equiv.Set.ofEq <| show _ '' (setOf _) = setOf _ by
+    ext r; obtain ⟨r, rfl⟩ := surjective_frobenius R p r
+    simp [expand_eq_zero (Fact.out : p.Prime).pos, (frobenius_inj R p).eq_iff, ← frobenius_def]
+
+theorem rootsExpandEquivRoots_apply (x) : (rootsExpandEquivRoots p f x).1 = x.1 ^ p := rfl
```

Comment:
```suggestion
theorem rootsExpandEquivRoots_apply (x) : (rootsExpandEquivRoots p f x : R) = x ^ p := rfl
```
Should this be a simp lemma?

Reply:
> Should this be a simp lemma?

I'm not sure. As a reference, [frobenius_def](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#frobenius_def) and [iterate_frobenius](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#iterate_frobenius) are not simp lemmas.

Reply:
Yes but this is the coercion of the application. After some more consideration, I think we should make this a simp lemma.